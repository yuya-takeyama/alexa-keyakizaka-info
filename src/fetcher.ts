import axios from 'axios';
import { JSDOM } from 'jsdom';
import { first, map, tail, trim } from 'lodash';
import * as moment from 'moment';
import * as striptags from 'striptags';

const BASE_URL = 'http://www.keyakizaka46.com';

const toURL = (path: string): string => {
  return `${BASE_URL}${path}`;
};

export interface Schedule {
  title?: string;
  time?: string;
  genre?: string;
  description?: string;
}

const normalize = (str: string | undefined | null): string | undefined => {
  return str ? trim(striptags(str)) || undefined : undefined;
};

const parseScheduleElement = (element: Element): Schedule | undefined => {
  const titleDescriptionElement = element.querySelector(
    '.box-detail_txt :last-child',
  );
  const genreElement = element.querySelector('.box-detail_genre');
  const timeElement = element.querySelector('.box-detail_txt .time');

  if (titleDescriptionElement) {
    const titleDescriptions = trim(
      striptags(titleDescriptionElement.innerHTML),
    ).split(/\n/);
    const title = first(titleDescriptions);
    const description = normalize(tail(titleDescriptions).join('\n'));

    return {
      title,
      genre: genreElement ? normalize(genreElement.innerHTML) : undefined,
      time: timeElement ? normalize(timeElement.innerHTML) : undefined,
      description,
    };
  } else {
    return undefined;
  }
};

const today = (): moment.Moment => moment.utc({ hour: 0 });

export const fetchSchedules = async (
  date: moment.Moment = today(),
): Promise<Schedule[]> => {
  return new Promise<Schedule[]>(async (resolve, reject) => {
    try {
      const res = await axios.get(
        toURL(`/s/k46o/media/list?dy=${date.format('YYYYMMDD')}`),
      );
      const scheduleElements = new JSDOM(
        res.data,
      ).window.document.querySelectorAll(
        '#schedule .box-schedule_inner .box-detail',
      );
      const schedules = map(scheduleElements, element =>
        parseScheduleElement(element),
      ).filter(schedule => schedule) as Schedule[];

      resolve(schedules);
    } catch (err) {
      reject(err);
    }
  });
};
