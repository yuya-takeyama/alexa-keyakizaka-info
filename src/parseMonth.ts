export type BirthdayParameter = BirthdayMonth | BirthdayDate;

export interface BirthdayMonth {
  type: 'month';
  year: number;
  month: number;
}

export interface BirthdayDate {
  type: 'date';
  year: number;
  month: number;
  day: number;
}

export default (str: string): BirthdayParameter => {
  const [year, month, day] = str.split('-');
  if (year && month) {
    if (day) {
      return {
        type: 'date',
        year: parseInt(year, 10),
        month: parseInt(month, 10),
        day: parseInt(day, 10),
      };
    } else {
      return {
        type: 'month',
        year: parseInt(year, 10),
        month: parseInt(month, 10),
      };
    }
  } else {
    throw new Error(`Invalid month|date format: ${str}`);
  }
};
