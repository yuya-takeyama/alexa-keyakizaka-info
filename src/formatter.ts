import { Moment } from "moment-timezone";
import { Schedule, ScheduleTime } from "./fetcher";

const formatTime = (time: ScheduleTime): string | undefined => {
  if (time.from || time.to) {
    return (time.from ? `${time.from}から` : '') +
    (time.to ? `${time.to}まで` : '');
  } else {
    return undefined;
  }
};

const formatSchedule = (schedule: Schedule): string => {
  return [formatTime(schedule.time), schedule.genre, schedule.title]
    .filter(elem => elem)
    .join(' ');
};

export const formatSchedules = (schedules: Schedule[], date: Moment): string => {
  const script = schedules.reduce((prev, schedule) => {
    return prev + formatSchedule(schedule) + '\n';
  }, '');
  return `${date.format('YYYY/MM/DD')}のスケジュールは${schedules.length}件です。\n${script}`;
}
