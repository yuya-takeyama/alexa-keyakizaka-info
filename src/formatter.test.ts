import * as moment from 'moment';
import { Schedule, Birthday } from './fetcher';
import { formatSchedules, formatBirthdays } from './formatter';
import { BirthdayMonth, BirthdayDate } from './parseMonth';

describe('#formatSchedules', () => {
  const date = moment('2017-12-31');

  describe('with no schedules', () => {
    it('returns a schedule script correctly', () => {
      expect(formatSchedules([], date)).toMatchSnapshot();
    });
  });

  describe('with a single schedule', () => {
    it('returns a schedule script correctly', () => {
      const schedules: Schedule[] = [
        {
          title: '歌番組に出演',
          genre: 'テレビ',
          time: {
            from: '19:00',
            to: '20:00',
          },
          description: undefined,
        },
      ];
      expect(formatSchedules(schedules, date)).toMatchSnapshot();
    });
  });

  describe('with multiple schedules', () => {
    it('returns a schedule script correctly', () => {
      const schedules: Schedule[] = [
        {
          title: '歌番組に出演',
          genre: 'テレビ',
          time: {
            from: '19:00',
            to: '20:00',
          },
          description: undefined,
        },
        {
          title: '平手友梨奈の誕生日',
          genre: '誕生日',
          time: {},
          description: '平手友梨奈の誕生日です。',
        },
        {
          title: '欅坂46 こちら有楽町星空放送局',
          genre: 'ラジオ',
          time: {
            from: '24:20',
            to: '24:40',
          },
          description: undefined,
        },
      ];
      expect(formatSchedules(schedules, date)).toMatchSnapshot();
    });
  });
});

describe('#formatBirthdays', () => {
  describe('with BirthdayMonth', () => {
    const time: BirthdayMonth = {
      type: 'month',
      year: 2017,
      month: 11,
    };

    describe('with no schedules', () => {
      it('returns a schedule script correctly', () => {
        expect(formatBirthdays([], time)).toMatchSnapshot();
      });
    });

    describe('with a single schedule', () => {
      it('returns a schedule script correctly', () => {
        const birthdays: Birthday[] = [
          {
            date: moment('2017-06-25'),
            name: '平手友梨奈',
          },
        ];
        expect(formatBirthdays(birthdays, time)).toMatchSnapshot();
      });
    });

    describe('with multiple schedules', () => {
      it('returns a schedule script correctly', () => {
        const schedules: Birthday[] = [
          {
            date: moment('2017-06-04'),
            name: '織田奈那',
          },
          {
            date: moment('2017-06-25'),
            name: '平手友梨奈',
          },
        ];
        expect(formatBirthdays(schedules, time)).toMatchSnapshot();
      });
    });
  });

  describe('with BirthdayDate', () => {
    const time: BirthdayDate = {
      type: 'date',
      year: 2017,
      month: 6,
      day: 25,
    };

    describe('with no schedules', () => {
      it('returns a schedule script correctly', () => {
        expect(formatBirthdays([], time)).toMatchSnapshot();
      });
    });

    describe('with a single schedule', () => {
      it('returns a schedule script correctly', () => {
        const birthdays: Birthday[] = [
          {
            date: moment('2017-06-25'),
            name: '平手友梨奈',
          },
        ];
        expect(formatBirthdays(birthdays, time)).toMatchSnapshot();
      });
    });

    describe('with multiple schedules', () => {
      it('returns a schedule script correctly', () => {
        const schedules: Birthday[] = [
          {
            date: moment('2017-06-25'),
            name: '平手友梨奈',
          },
          {
            date: moment('2017-06-25'),
            name: '平手友梨奈2',
          },
        ];
        expect(formatBirthdays(schedules, time)).toMatchSnapshot();
      });
    });
  });
});
