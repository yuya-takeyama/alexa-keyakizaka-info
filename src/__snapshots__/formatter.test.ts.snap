// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#formatBirthdays with BirthdayDate with a single schedule returns a schedule script correctly 1`] = `
"6月25日が誕生日のメンバーは1人です。
平手友梨奈
"
`;

exports[`#formatBirthdays with BirthdayDate with multiple schedules returns a schedule script correctly 1`] = `
"6月25日が誕生日のメンバーは2人です。
平手友梨奈
平手友梨奈2
"
`;

exports[`#formatBirthdays with BirthdayDate with no schedules returns a schedule script correctly 1`] = `
"6月25日が誕生日のメンバーは0人です。
"
`;

exports[`#formatBirthdays with BirthdayMonth with a single schedule returns a schedule script correctly 1`] = `
"11月が誕生日のメンバーは1人です。
25日、平手友梨奈
"
`;

exports[`#formatBirthdays with BirthdayMonth with multiple schedules returns a schedule script correctly 1`] = `
"11月が誕生日のメンバーは2人です。
04日、織田奈那
25日、平手友梨奈
"
`;

exports[`#formatBirthdays with BirthdayMonth with no schedules returns a schedule script correctly 1`] = `
"11月が誕生日のメンバーは0人です。
"
`;

exports[`#formatSchedules with a single schedule returns a schedule script correctly 1`] = `
"2017/12/31のスケジュールは1件です。
19:00から20:00まで テレビ 歌番組に出演
"
`;

exports[`#formatSchedules with multiple schedules returns a schedule script correctly 1`] = `
"2017/12/31のスケジュールは3件です。
19:00から20:00まで テレビ 歌番組に出演
誕生日 平手友梨奈の誕生日
24:20から24:40まで ラジオ 欅坂46 こちら有楽町星空放送局
"
`;

exports[`#formatSchedules with no schedules returns a schedule script correctly 1`] = `
"2017/12/31のスケジュールは0件です。
"
`;
